### PR TITLE
Ported CA2227 and added Shared Lib

### PR DESCRIPTION
**Pull request by heejaechang**

### CODE REVIEW

**Comment by heejaechang** [comment]
@srivatsn @mavasani @lgolding @jinujoseph @balajikris can you take a look?

most of changes are just added using for shared lib. and actual change are actually in just 2 files.

File: PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.cs [hunk 1]
```diff
@@ -1,38 +1,101 @@
...
+                                                                    isEnabledByDefault: true,
+                                                                    description: s_localizableDescription,
+                                                                    helpLinkUri: "https://msdn.microsoft.com/library/ms182327.aspx",
+                                                                    customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get
+            {
+                return ImmutableArray.Create(Rule);
```
Comment (genlu) [inline PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.cs]: Why not using expression body syntax?

File: PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.cs [hunk 2]
```diff
@@ -1,38 +1,101 @@
...
+                    context.RegisterSymbolAction(c => AnalyzeSymbol(c, iCollectionType, arrayType), SymbolKind.Property);
+                });
+        }
+
+        public void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol iCollectionType, INamedTypeSymbol arrayType)
+        {
+            var property = context.Symbol as IPropertySymbol;
+            if (property == null)
+            {
+                return;
```
Comment (genlu) [inline PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.cs]: This is not necessary, a direct cast should be fine.

File: PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.Fixer.cs [hunk 3]
```diff
@@ -1,36 +0,0 @@
...
-        }
-
-        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
-        {                              
-            // This is to get rid of warning CS1998, please remove when implementing this analyzer
-            await new Task(() => { });
-            throw new NotImplementedException();
-        }
-    }
-}
```
Comment (genlu) [inline PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.Fixer.cs]: I actually think a CodeFix would work well for this rule.

File: PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.cs (hunk 1 above)
Comment (heejaechang) [inline PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.cs]: sure

File: PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.cs (hunk 2 above)
Comment (heejaechang) [inline PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.cs]: sure

File: PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.Fixer.cs (hunk 3 above)
Comment (heejaechang) [inline PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.Fixer.cs]: just removing setter is not right fix though. recommended way is adding method like Clear or SetXXXX. but looks like not it is not one pattern to fix but depends on code.

**Comment by genlu** [comment]
Another thing I forgot to mention. Is it possible to use `git mv` for all the shared utility code instead of creating new files?  This way we can see in history where are those code come from. But if it is too much work, this is fine too.

File: PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.Fixer.cs (hunk 3 above)
Comment (genlu) [inline PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.Fixer.cs]: This is non-blocking though. You can just create an issue for CodeFix and add it later.

**Comment by heejaechang** [comment]
@genlu I think for those 2 files, probably okay to leave like I have now? :)

**Comment by genlu** [comment]
@heejaechang Yeah, sure.

File: PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.Fixer.cs (hunk 3 above)
Comment (genlu) [inline PortFxCop/src/Microsoft.ApiDesignGuidelines.Analyzers/Core/CollectionPropertiesShouldBeReadOnly.Fixer.cs]: I see, if this is the case then you can ignore my comment. I looked at the sample code on MSDN and thought we just need to remove setter and let user use Clear and AddRange methods directly on the collection.

**Comment by genlu** [comment]
:+1:

**Pull request merged**